### PR TITLE
added changes for better build ordering when SQL expressions are pres…

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes

Improvement to reduce number of times explicit baseColumn is needed to be specified when a column 
definition has references to previous columns inside of the SQL expression clause. 

It uses a simple identifier parser to determine when there are potential references to earlier columns and adjusts the separation of phases for column building

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Change to tutorials, tests or examples
- [ ] Non code change (readme, images or other non-code assets)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [ ] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [ ] Submission does not increase alerts or messages from LGTM

## Further comments

Each phase of column generation produces a select statement to generate the columns. When the SQL expression refers to other columns, they must have been generated in a previous select statement. 

This adjusts the assignment of column generators to select statements.

### CODE REVIEW

File: dbldatagen/schema_parser.py [hunk 1]
```diff
@@ -132,7 +132,7 @@ def getTypeDefinitionParser(cls):
                 pp.delimitedList(pp.Group(ident + pp.Optional(colon) + pp.Group(type_expr)))) + r_angle
 
             # try to capture invalid type name for better error reporting
-            invalid_type = pp.Word(pp.alphas, pp.alphanums+"_", as_keyword=True)
+            invalid_type = pp.Word(pp.alphas, pp.alphanums+"_", asKeyword=True)
```
Comment (howardwu-db) [inline dbldatagen/schema_parser.py:135]: they are probably getting rid of asKeyword "asKeyword and useRegex are retained for pre-PEP8 compatibility, but will be removed in a future release" https://pyparsing-docs.readthedocs.io/en/latest/pyparsing.html so the previously written version of this probably better

File: dbldatagen/schema_parser.py [hunk 2]
```diff
@@ -260,6 +260,67 @@ def columnTypeFromString(cls, type_string):
 
         return type_construct
 
+    @classmethod
+    def _cleanseSQL(cls, sql_string):
+        """ Cleanse sql string removing strings so that they are not considered as part of potential column
```
Comment (howardwu-db) [inline dbldatagen/schema_parser.py:265]: Cleanse sql strings **BY** removing strings

File: dbldatagen/schema_parser.py [hunk 3]
```diff
@@ -260,6 +260,67 @@ def columnTypeFromString(cls, type_string):
...
+        quoted_ident = pp.QuotedString(quoteChar="`", escQuote="``")
+
+        stringForm1 = pp.Literal('r') + pp.QuotedString(quoteChar="'")
+        stringForm2 = pp.Literal('r') + pp.QuotedString(quoteChar='"')
+        stringForm3 = pp.QuotedString(quoteChar="'", escQuote=r"\'")
+        stringForm4 = pp.QuotedString(quoteChar='"', escQuote=r'\"')
+        stringForm = stringForm1 ^ stringForm2 ^ stringForm3 ^ stringForm4
+        stringForm.set_parse_action(lambda s, loc, toks: "' '")
+
+        parser = quoted_ident ^ stringForm
```
Comment (howardwu-db) [inline dbldatagen/schema_parser.py:284]: does quoted_ident also need to have .set_parse_action applied to it for the transform_string to work for that?

File: dbldatagen/utils.py [hunk 4]
```diff
@@ -214,3 +215,57 @@ def parse_time_interval(spec):
...
+    x = ['id', 'city_name', 'id', 'city_id', 'city_pop', 'id', 'city_id', 'city_pop','city_id', 'city_pop','id']
+    splitListOnCondition(x, lambda el: el == 'id')
+
+
+    result:
+    `[['id'], ['city_name'], ['id'], ['city_id', 'city_pop'],
+     ['id'], ['city_id', 'city_pop', 'city_id', 'city_pop'], ['id']]`
+
+    :arg lst: list of items to perform condition matches against
+    :arg cond: lambda function or function taking single argument and returning True or Fals
```
Comment (howardwu-db) [inline dbldatagen/utils.py:239]: typo for false here

File: dbldatagen/utils.py [hunk 5]
```diff
@@ -214,3 +215,57 @@ def parse_time_interval(spec):
...
+    :returns: list of sublists
+    """
+
+    def match_condition(matchList, matchFn):
+        """Return first index of element of list matching condition"""
+        if matchList is None or len(matchList) == 0:
```
Comment (howardwu-db) [inline dbldatagen/utils.py:245]: this is redundant due to the if statement on line 257

File: docs/source/generating_column_data.rst [hunk 6]
```diff
@@ -107,13 +107,29 @@ This performs the following actions:
...
 
-- If a column is referred to in an expression, the ``baseColumn`` attribute must be defined with a dependency
+- If a column is referred to in an expression, the ``baseColumn`` attribute may need to be defined with a dependency
   on that column
 - If a column uses a base column with a restricted range of values then it is possible that the column
   will not generate the full range of values in the column generation spec
 - If the base column is of type ``boolean`` or some other restricted range type, computations on that base value
   may not produce the expected range of values
-- If base column is not specified, you may see errors reporting that the column in an expression does not exist
+- If base column is not specified, you may see errors reporting that the column in an expression does not exist. T
```
Comment (howardwu-db) [inline docs/source/generating_column_data.rst:124]: Typo on the extra T here

File: dbldatagen/schema_parser.py (hunk 1 above)
Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py:135]: i checked and it does not work in some versions of the Databricks runtime while the use of asKeyword does. 

They may change this in future but for maximum compatibility with existing versions - we need to use `asKeyword`

File: dbldatagen/schema_parser.py (hunk 2 above)
Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py:265]: added clarification : cleanse SQL string by removing string literals ...

File: dbldatagen/schema_parser.py (hunk 3 above)
Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py:284]: no it does not. 

What we're looking to do is determine which parts of a sql expression could be field references.

For the expression 'concat('hello', field1, `field 2`)' - field1 and `field 2` could be field references. 

'hello' is not

Comment (ronanstokes-db) [reply dbldatagen/schema_parser.py:284]: There are some rare cases where a field reference could be inside a literal string - but user can still override with `baseColumn=...`

Goal is to reduce amount of times user has to explicitly declare a reference.

File: dbldatagen/utils.py (hunk 5 above)
Comment (ronanstokes-db) [reply dbldatagen/utils.py:245]: Not quite correct - consider case where len(matchlist) == 0

Outer if does not check for it. 

In any event want inner `match_condition` to guard against empty list 

We could drop outer check but cant take len(None)

File: docs/source/generating_column_data.rst (hunk 6 above)
Comment (ronanstokes-db) [reply docs/source/generating_column_data.rst:124]: fixed

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**